Classes/EventListener/Order/Stock/HandleStock.php

```diff
@@ -11,77 +11,47 @@
-            if ($cartProduct->getProductType() === 'CartProducts') {
-                $productConnection = GeneralUtility::makeInstance(ConnectionPool::class)
-                    ->getConnectionForTable('tx_cartproducts_domain_model_product_product');
-                $productQueryBuilder = $productConnection->createQueryBuilder();
-
-                $product = $productQueryBuilder
-                    ->select('uid', 'handle_stock', 'handle_stock_in_variants')
-                    ->from('tx_cartproducts_domain_model_product_product')->where($productQueryBuilder->expr()->eq('uid', $productQueryBuilder->createNamedParameter($cartProduct->getProductId(), \PDO::PARAM_INT)))->executeQuery()->fetchAssociative();
+            if ($cartProduct->getProductType() !== 'CartProducts' || $cartProduct->isHandleStock() === false) {
+                continue;
+            }
 
-                if ($product['handle_stock']) {
-                    if ($product['handle_stock_in_variants']) {
-                        $this->handleStockInBeVariant($cartProduct);
-                    } else {
-                        $this->handleStockInProduct($cartProduct);
-                    }
+            if ($cartProduct->isHandleStockInVariants()) {
+                foreach ($cartProduct->getBeVariants() as $cartProductBeVariant) {
+                    $this->handleStockInBeVariant($cartProductBeVariant);
                 }
+            } else {
+                $this->handleStockInProduct($cartProduct);
             }
         }
     }
 
-    protected function handleStockInProduct(CartProduct $cartProduct): void
+    private function handleStockInProduct(CartProduct $cartProduct): void
     {
-        $productConnection = GeneralUtility::makeInstance(ConnectionPool::class)
-            ->getConnectionForTable('tx_cartproducts_domain_model_product_product');
-        $productQueryBuilder = $productConnection->createQueryBuilder();
-
-        $product = $productQueryBuilder
-            ->select('stock')
-            ->from('tx_cartproducts_domain_model_product_product')->where($productQueryBuilder->expr()->eq('uid', $productQueryBuilder->createNamedParameter($cartProduct->getProductId(), \PDO::PARAM_INT)))->executeQuery()->fetchAssociative();
-
-        $productQueryBuilder
-            ->update('tx_cartproducts_domain_model_product_product')
-            ->where(
-                $productQueryBuilder->expr()->eq('uid', $productQueryBuilder->createNamedParameter($cartProduct->getProductId(), \PDO::PARAM_INT))
-            )
-            ->orWhere(
-                $productQueryBuilder->expr()->eq('l10n_parent', $productQueryBuilder->createNamedParameter($cartProduct->getProductId(), \PDO::PARAM_INT))
-            )->set('stock', $product['stock'] - $cartProduct->getQuantity())->executeStatement();
+        $quantityToAdd = -1 * $cartProduct->getQuantity();
+        $this->productRepository->addQuantityToStock($cartProduct->getProductId(), $quantityToAdd);
```

Comment:
My first method name was `updateStock()` but this was also misleading because one doesn't know what `$quantity` is. The quantity to add, to substract ot to set. I decide to rename it to `addQuantityToStock()`. My intention was, that this method can be used by payment providers to update (add) the stock on order cancelation. I will provide another method in the DoctrineRepositories, which than internally multiply the amount and call the `addQuantityToStock()` method.

Reply:
I added an method to the repository in 9907132.